### PR TITLE
Add experimental pushdown for dot notation

### LINKED ISSUE #2069: Re-add pushdown for dot notation
See #2055 and https://github.com/FerretDB/FerretDB/pull/2068#issuecomment-1448131841.

Let's add it back for PostgreSQL only, in simplest cases, and under a feature flag.

### PR DESCRIPTION
**Pull request by noisersup**
# Description

Closes #2069.

## Readiness checklist

- [x] I added/updated unit tests (and they pass).
- [ ] I added/updated integration/compatibility tests (and they pass).
- [ ] I added/updated comments and checked rendering.
- [x] I made spot refactorings.
- [ ] I updated user documentation.
- [ ] I ran `task all`, and it passed.
- [x] I ensured that PR title is good enough for the changelog.
- [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Milestone (`Next`), Labels, Project and project's Sprint fields.
- [x] I marked all done items in this checklist.

### CODE REVIEW

**Review by henvic — APPROVED** [review]
It looks good to me.

File: internal/backends/postgresql/query.go [hunk 1]
```diff
@@ -119,7 +122,9 @@ func prepareWhereClause(p *metadata.Placeholder, sqlFilters *types.Document) (st
 			// Handle dot notation.
 			// TODO https://github.com/FerretDB/FerretDB/issues/2069
 			if path.Len() > 1 {
-				continue
+				keyOperator = "#>"
+				key = path.Slice() // '{v,foo}'
+				//	continue
```
Comment (chilagrow) [inline internal/backends/postgresql/query.go:127]: Let's remove commented code
```suggestion
```

File: internal/handler/msg_aggregate.go [hunk 2]
```diff
@@ -281,6 +282,18 @@ func (h *Handler) MsgAggregate(ctx context.Context, msg *wire.OpMsg) (*wire.OpMs
 			qp.Filter = filter
 		}
 
+		if !h.EnableExperimentalPushdown && filter != nil {
```
Comment (chilagrow) [inline internal/handler/msg_aggregate.go:285]: If `h.DisablePushdown == true`, is  `h.EnableExperimentalPushdown == true` intended to override it by enabling pushdown?

**Review by chilagrow — COMMENTED** [review]
Great work! Was this tested manually? I have one question :hugs:

File: cmd/ferretdb/main.go [hunk 3]
```diff
@@ -92,7 +92,8 @@ var cli struct {
 	Test struct {
 		RecordsDir string `default:"" help:"Testing: directory for record files."`
 
-		DisablePushdown bool `default:"false" help:"Experimental: disable pushdown."`
+		DisablePushdown            bool `default:"false" help:"Experimental: disable pushdown."`
+		EnableExperimentalPushdown bool `default:"false" help:"Experimental: enable experimental pushdown."`
```
Comment (AlekSi) [inline cmd/ferretdb/main.go:96]: I don't think we need both "test" and "experimental" in the flag name, and say that experimental flag is experimental in the help text.

```suggestion
		EnableNestedPushdown bool `default:"false" help:"Experimental: enable pushdown for nested fields."`
```

File: internal/backends/postgresql/query.go [hunk 4]
```diff
@@ -119,7 +122,9 @@ func prepareWhereClause(p *metadata.Placeholder, sqlFilters *types.Document) (st
 			// Handle dot notation.
 			// TODO https://github.com/FerretDB/FerretDB/issues/2069
```
Comment (AlekSi) [inline internal/backends/postgresql/query.go:123]:
```suggestion
```

File: internal/backends/postgresql/query.go [hunk 5]
```diff
@@ -103,6 +103,9 @@ func prepareWhereClause(p *metadata.Placeholder, sqlFilters *types.Document) (st
 			return "", nil, lazyerrors.Error(err)
 		}
 
+		keyOperator := "->"   // keyOperator is the operator that is used to access the field. (->/#>)
+		var key any = rootKey // key can be either a string '"v"' or PostgreSQL path '{v,foo}'
```
Comment (AlekSi) [inline internal/backends/postgresql/query.go:107]: What is missing is a comment explaining why we can't always use `#>` and `{v}` even with a single field

File: internal/handler/msg_aggregate.go (hunk 2 above)
Comment (AlekSi) [reply internal/handler/msg_aggregate.go:285]: I would say we should check both are not true at the same time in the main.go

**Review by chilagrow — APPROVED** [review]
Looks good to me!

**Comment by AlekSi** [comment]
Waiting for feedback

**Review by AlekSi — APPROVED** [review]

**Pull request merged**